### PR TITLE
8276167: VirtualFlow.scrollToTop doesn't scroll to the top of the last element

### PR DESCRIPTION
**Pull request by eduardsdv**
Fix VirtualFlow.scrollToTop(int) doesn't scroll to the top of the last element but to the bottom of the last element.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276167](https://bugs.openjdk.java.net/browse/JDK-8276167): VirtualFlow.scrollToTop doesn't scroll to the top of the last element

### Reviewers
 * [Ajit Ghaisas](https://openjdk.java.net/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Johan Vos](https://openjdk.java.net/census#jvos) (@johanvos - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/656/head:pull/656` \
`$ git checkout pull/656`

Update a local copy of the PR: \
`$ git checkout pull/656` \
`$ git pull https://git.openjdk.java.net/jfx pull/656/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 656`

View PR using the GUI difftool: \
`$ git pr show -t 656`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/656.diff">https://git.openjdk.java.net/jfx/pull/656.diff</a>

</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
This might be related to [JDK-8276553](https://bugs.openjdk.java.net/browse/JDK-8276553) / PR #683.

After that one is reviewed and integrated, this bug needs to be retested to see if it is still a problem. If it is, then the fix may need to be adjusted.

**Comment by kevinrushforth** [comment]
Two global comments on this review:

1. Can you enable GitHub Actions testing for your repo as [described here](https://github.com/openjdk/jfx/pull/656/checks?check_run_id=4046432049)?
2. You need to change the title of this PR to match that of the JBS issue.

**Comment by eduardsdv** [comment]
I tested it again. Without the changes in this PR the bug is still there. 
If the item is larger than the viewport, the VirtualFlow.scrollToTop(int) scrolls to the end instead of to top of the item.

**Comment by kleopatra** [comment]
> I tested it again. Without the changes in this PR the bug is still there. **If the item is larger than the viewport**, the VirtualFlow.scrollToTop(int) scrolls to the end instead of to top of the item.

ahhh .. finally I understand what you mean - bolding of the quote ;) Might be obvious for everybody except me .. but the bug report is missing that crucial information ...

**Comment by johanvos** [comment]
I agree with @kleopatra . It would be good to have this info in JBS, as well as with a small sample app.

**Comment by johanvos** [comment]
@eduardsdv This would be a good fix, but can you add a test that fails before and succeeds after this patch? Let me know if you need help with that. Thanks!

**Comment by eduardsdv** [comment]
I added a junit which fails without patch and succeeds with it.

**Comment by eduardsdv** [comment]
Please review.

**Review by aghaisas — APPROVED** [review]
Fix looks good.
I verified that the tests fail without fix and pass with it.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 1]
```diff
@@ -1581,7 +1581,7 @@ private boolean tryScrollOneCell(int targetIndex, boolean downOrRight) {
     public void scrollToTop(int index) {
         boolean posSet = false;
 
-        if (index >= getCellCount() - 1) {
+        if (index > getCellCount() - 1) {
```
Comment (johanvos) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:1584]: I agree that the previous code violates the contract that the flow should show the specified cell aligned to the top.

I wonder though if this test is needed at all. What is the goal of providing an index that is larger than getCellCount() -1?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewKeyInputTest.java [hunk 2]
```diff
@@ -1841,6 +1841,7 @@ private boolean isAnchor(int index) {
             listView.getItems().add("Row " + i);
         }
         listView.setPrefHeight(130); // roughly room for four rows
+        Toolkit.getToolkit().firePulse();
```
Comment (johanvos) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewKeyInputTest.java:1844]: Why is this needed here?

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 1 above)
Comment (eduardsdv) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:1584]: This part was originally copied from the 'VirtualContainerBase'. There was this comment: _special-case the 'greater than row count' condition to have it be perfectly at position 1_.

Maybe to avoid throwing the IndexOutOfBoundsException or simply to scroll to the end.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewKeyInputTest.java (hunk 2 above)
Comment (eduardsdv) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewKeyInputTest.java:1844]: Without this line the test fails in the line 1865 with the message 'expected:<7> but was:<8>'.
I think this is because in the line 1850 'listView.scrollTo(99)' is executed, which now does not set the position to 1.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 1 above)
Comment (johanvos) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:1584]: I am ok with this behavior, but then it should be specified in the javadoc for `@parameter index` that the value can be greater than row count, and that in that case the view will be positioned at the very end.
@kevinrushforth do you have an opinion on this?
This is unrelated to the current fix, so that should not stop this PR.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewKeyInputTest.java (hunk 2 above)
Comment (johanvos) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewKeyInputTest.java:1844]: That is very well possible indeed. Since we estimate the size of the different cells, it is hard to deterministically put values on the counts. Therefor, I changed a number of tests from `equals` to `<` . I would personally suggest not adding the `pulse` here, but rather change the test in line 1865 with `... < 10 ` to make sure that there is no abnormal amount of processing. (and maybe also `> 0 ` to make sure that the updateItem is at least counted once).

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 1 above)
Comment (kevinrushforth) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:1584]: This seems reasonable to me, too. And I agree that it should be specified via a follow-up issue, which will need a CSR.

**Review by johanvos — APPROVED** [review]

**Review by aghaisas — APPROVED** [review]
